Pet Buffs should be saved & SpellEffect::Threat should be displayed in combatlog

dagochen opened the pull request:
The first is commit is for pet buffs. If you dismiss a buffed pet and call it again, the buffs are gone. They should be saved.

The second commit is for threat-spells to appear in combatlog. Thats important for addons like KTM Threatmeter, because it parses the combatlog messages on the client and calculates the threat value. In KTM it says that this kind of message need to be like this: "Dagochen performs Faint on ". But I didn't manage to get that "on " working. It currently only says "Dagochen performs Faint."

dagochen commented:
@Phatcat Pet related!

And the threat thing needs some improvements to put the target into combatlog

Phatcat commented:
I don't quite understand the purpose of the first commit; what exactly is the issue with pets and auras?
Also, what exactly are you trying to acomplish with the second commit? I am not strong on logs, there are so many different logging/message systems in our system.

I am not saying any of this is wrong at all, I just don't understand it clearly.

When I understand it I'll be able to review and test and maybe even help you improve it further =)

dagochen commented:
The first is commit is for pet buffs. If you dismiss a buffed pet and call it again, the buffs are gone. They should be saved.

The second commit is for threat-spells to appear in combatlog. Thats important for addons like KTM Threatmeter, because it parses the combatlog messages on the client and calculates the threat value. In KTM it says that this kind of message need to be like this: "Dagochen performs Faint on <Target>". But I didn't manage to get that "on <Target>" working. It currently only says "Dagochen performs Faint."

Phatcat commented:
> we only want to save auras, which dont have 1:1 relation. Otherwise a…
…n aura with 1:1 relation could be placed on a pet, the pet gets saved to db and then someone else would get the aura as well."

How could we ever end up saving an aura that isn't related to the pet, though?
also That's gotta be the most confusing Pr title then; shouldn't it be more like "Save Pet auras on dismiss"?

Alright, now I understand what you want.. well, why don't you add them to the log, then? - do you not have the target available in the function that does the logging?
I am not expert in this but if you have access to the unit, then you should be able to get the name string and be able to pass it along like any other string; all units/objects in the game has names if I am not mistaken.

dagochen commented:
For some spells it is working if you add the targetguid to the packet which is sent to the clients. For Faint it does not. I hoped someone might have an idea how to get this working.

Phatcat commented:
What's wrong with sending [this unit](https://github.com/cmangos/mangos-classic/pull/167/files#diff-b4db11cbf2813cd24b3c165885fc08a7R3357) along, by the way?

Phatcat commented on src/game/Spell.cpp:
```diff
@@ -3375,6 +3375,13 @@ void Spell::SendLogExecute()
+                case SPELL_EFFECT_DISPEL:
+                case SPELL_EFFECT_THREAT:
+                    if (Unit* unit = m_targets.getUnitTarget())
+                        data << unit->GetPackGUID();
```
if (target = m_targets.getUnitTarget())
data << target->GetPackGUID();

dagochen commented:
Its the way it is done in every case of the switch statement in that method. I did not thought about that, just did it the same way to be consistent.

Phatcat commented:
You're right, it is done like in all the other cases; it doesn't make it less of a waste, though.
What is the point of the pointer 'target'?

I am just now reading the file and... Yeah... You are doing the right thing, it's
https://github.com/cmangos/mangos-classic/pull/167/files#diff-b4db11cbf2813cd24b3c165885fc08a7R3357 that appears to be bloat, as the variable pointer 'target' doesn't seem to be used anywhere at all...

dagochen commented:
With if (target = m_targets.getUnitTarget()) you change the pointer to null if there is no unittarget. In that case target should be caster (as mentioned in the first line where target gets initialized). I think its not that important for this case. I am more interested in getting it work as intended, so that addons like KTM can track threat modifications by Feint, Disengage and so on as well. Would make the life of every Rogue and Feral Druid so much easier.

Phatcat commented:
You're right, what I'm talking about has little to no impact on the server itself; it is just bad design, to set a variable pointer and never use it. Got nothing to do with this PR, I just made a note of it since your PR made me notice it.

killerwife commented:
tabs

dagochen commented:
Thanks @killerwife

Phatcat commented:
Looks good to me, can't see how this could regress on anything; going to test and merge in a few days time.

Phatcat commented:
You mind squashing / fixing up these commits? (probably into 2, one for auras, one for logging)

dagochen commented:
I dont understand what you want me to do. It's already 2 (or 3) commits. The first is for Pet Auras and 2nd+3rd for the logging thing.

Phatcat commented:
No, there are 4 commits, also the logging don't need 2 commits, just needs 1.

dagochen commented:
Still dont know what to do....sorry.

Phatcat commented:
You on Windows?

dagochen commented:
Yep

Phatcat commented:
If you are on linux or not using the github client then you already know how to open a git command line tool, such as git bash. Open that and navigate to the branch you have PR'd from.

However, if you are using the github client then, again, navigate to the branch you are PR'ing from within the client and then press the gear icon in the upper right corner and press `Open in Git Shell`

first enter this command
`git rebase -i head~4`

This will open a text document, here you will see a somewhat simple guide written in the comments, and above it 4 lines, the 4 latest commits on the branch; what you want to do is leave the first one be, the top one, the second one you also leave be, but the third line you replace ´pick´ with either `s` or `f` depending if you want to save some of the message from that commit or not, the forth line you replace `pick` with `f`

When you are done then save the document and close it.

If you chose `s` and not `f` for the third line then another text document will open; here you have the chance to touch up on the commit messages if you are not happy with them.

When you are done, save and close the document.

When you are done with squashing/fixing up, you do a rebase, just to make sure your code is still able to merge with current master;

`git rebase cmangos/master`

if it tells you it does not know a cmangos repository, then you need to add it to remote, you do this by;

`git remote add cmangos https://github.com/cmangos/mangos-classic.git`

followed by a

`git fetch cmangos`

Now you should be able to do a

`git rebase cmangos/master`

When all of that is done you force push your local changes to remote; this will update the PR. The reason you have to force push it is because you have altered commits that have already been pushed; had you only altered commits you had not pushed to remote yet at all, a force would not be needed.

`git push -f`

dagochen commented:
git status
HEAD detached from origin/dev

git push -f
fatal: You are not currently on a branch.

dago@dago-nb MINGW64 /d/Mangos/mangos ((7d31d8c...))

Now everything is broken...thanks :(

Phatcat commented:
Hey man, I'm sorry that something messed up in the process; I can promise you that it should not have happened if things were done properly. as I can see, nothing on this pr has changed, which kind of indicates that you were not on the right branch; the one you have PR'd from.

Again, none of the commands I've asked you to do are destructive in the sense that they remove code, if you have somehow done this on another branch which you used for osmething else, the code work is still in the files, the commits may just be a bit messed now.

If you hit me up on IRC I'll be able to help you in real time; nothing too bad could have happened really. Nothing unfixable at least.

dagochen commented:
Well nothing works anymore, cant push/pull/commit anything. I am out....

dagochen commented:
You can still see the commits. Go take whatever you need from then, fix it yourself.

Phatcat commented:
Okay man, I don't know how I could have offended you by trying to help you. Suit yourself.

Phatcat commented:
Yeah, you just made it more difficult contributing you since you deleted the branch, though. I could have taken over had you maybe said you didn't care to learn to squash/fixup - you were the one who asked for how to and now behave like I've insulted you. You think I'm wasting your time here or something cause it sure seems like you just wasted mine?

dagochen commented:
Nah its not that, I really appreciate the help. But I am just super-annoyed of this commandlineshit which almost NEVER works as intended. I have no knowledge about it and since there are only cryptographic error messages like "fatal: could not do this and that" I am so frustrated about it that I dont want to do anything with that anymore. I am sorry for the rage...wasnt against you for sure. Its just we are living in 2016 and there should be no need for that deep commandline knowledge anymore. I use to do everything with tortoisegit, which works really fine for my needs. Unfortunatly I could not find a squash/fixup button^^ And since I commit and fix things daily, this 4 commits are not even the "last 4 commits" in my branch. There is about 20 more. Propably all of them could be useful for this project....but it seems you guys are too busy already.

Phatcat commented:
Yeah, I'm not giving up on you yet, then ;)
https://1drv.ms/f/s!At-talbyMKRO-m-N5SVokutSKv7U

Phatcat commented:
A few things I think you've misunderstood;
Branch is not the same as fork; you have 1 fork of the repo, which you then branch out into as many as you need to work with; in your very case you have 3 branches; dev, master and 1 more I don't remember the name of, but you know. the **branch** you have PR'd this from has only 4 commits, these 4 commits. The other commits you are talking about you have on your master branch and on your other branch I don't remember the name of.
Yes, we are busy, but some of us are less busy than others, and I've said this before and I'll say this again; anything that I am tagged in I will look over, and review/test upon request. I cannot promise I'll merge your work, but I'll do anything in my power to help it along at least. No change is too insignifigant or uninteresting for me to look over either.

dagochen commented:
I get this error without changing/opening/doing anything, right after pulling from origin:
        modified:   dep/tbb/build/vs100project/makefile.sln
        modified:   dep/tbb/build/vs100project/tbb.vcxproj
        modified:   dep/tbb/build/vs100project/tbbmalloc.vcxproj
        modified:   dep/tbb/build/vs100project/tbbmalloc_proxy.vcxproj
        modified:   dep/tbb/build/vsproject/makefile.sln
        modified:   dep/tbb/build/vsproject/tbb.vcproj
        modified:   dep/tbb/build/vsproject/tbbmalloc.vcproj
        modified:   dep/tbb/build/vsproject/tbbmalloc_proxy.vcproj

dagochen commented:
I think I made it even worse :(

dagochen commented:
Is there anything I can do to get my repo back as it was BEFORE i tried squash....?

dagochen commented:
Its broken as fuck!

cyberium commented:
yes but no worries :)
come to irc if you need help its easyier

Muehe commented:
With a commit history like it shows on your repository, it might be easier to delete it, fork again, and then add your commits to branches other than master. The last part could be a bit tricky. You would have to either use `git cherry-pick`, which can be a bit confusing at first, or you just take the updated files from the backup of your old repository, paste them in the new one and write new commit messages.

dagochen commented:
I have been playing a little bit around it seems to be working now (not sure why or how, but who gives a fu**?) :D

I will create new PR now.
